src/test/index.spec.ts

```diff
@@ -1955,4 +2053,144 @@ test.suite('ts-node', (test) => {
       });
     }
   });
+
+  test.suite('top level await', (test) => {
+    const compilerOptions = {
+      target: semver.gte(ts.version, '3.0.1')
+        ? 'es2018'
+        : // TODO should this be kept if tla is not supported on TS <3.8.0?
+          // TS 2.7 is using polyfill for async interator even though they
+          // were added in es2018
+          'esnext',
+    };
+    function executeInTlaRepl(input: string, waitMs = 1000) {
+      return executeInRepl(
+        input
+          .split('\n')
+          .map((line) => line.trim())
+          .join(''),
+        {
+          waitMs,
+          createServiceOpts: {
+            experimentalReplAwait: true,
+            compilerOptions,
+          },
+          startOptions: { useGlobal: false },
+        }
+      );
+    }
+
+    if (semver.gte(ts.version, '3.8.0')) {
+      // Serial because it's timing-sensitive
+      test.serial('should allow evaluating top level await', async () => {
+        const script = `
+        export {}; /* TODO remove this line; implement automatic export{}; in the REPL */
+        const x: number = await new Promise((r) => r(1));
+        for await (const x of [1,2,3]) { console.log(x) };
+        for (const x of ['a', 'b']) { await x; console.log(x) };
+        class Foo {}; await 1;
+        function Bar() {}; await 2;
+        const {y} = await ({y: 2});
+        const [z] = await [3];
+        x + y + z;
+      `;
+
+        const { stdout, stderr } = await executeInTlaRepl(script);
+        expect(stderr).to.equal('');
+        expect(stdout).to.equal('> 1\n2\n3\na\nb\n6\n> ');
+      });
+
+      // Serial because it's timing-sensitive
+      test.serial(
+        'should wait until promise is settled when awaiting at top level',
+        async () => {
+          const awaitMs = 500;
+          const script = `
+          const startTime = new Date().getTime();
+          await new Promise((r) => setTimeout(() => r(1), ${awaitMs}));
+          const endTime = new Date().getTime();
+          endTime - startTime;
+        `;
+          const { stdout, stderr } = await executeInTlaRepl(script, 2500);
+
+          expect(stderr).to.equal('');
+
+          const ellapsedTime = Number(
+            stdout.split('\n')[0].replace('> ', '').trim()
+          );
+          expect(ellapsedTime).to.be.gte(awaitMs - 50);
```

Comment:
How could the promise resolve faster than 500ms?

Reply:
It resolves in 499 sometimes, making the tests fail.

Reply:
That surely is an interesting case, didn't know a timeout could finish before the given time, always though it must `>=`.